### PR TITLE
Support TLS when using the Mysql driver

### PR DESCRIPTION
**Pull request by surajvshankar**
## Summary
Extended xk6-sql to support mysql connection over TLS.

## Details
The mysql driver already supports TLS. These are additions to make xk6-sql TLS aware; such that, the TLS settings can be passed to the underlying drivers.  

> [!NOTE]  
> 1. No changes are required for the existing scripts. The `open` interface is retained as-is.
> 2. TLS is registered only if `enableTLS: true` is passed to `loadTLS` 

## Testing Done
```
./k6 run examples/mysql_secure_test.js -v
DEBU[0000] Logger format: TEXT                          
DEBU[0000] k6 version: v0.51.0 (commit/a3010c8ad2-dirty, go1.22.6, darwin/arm64) 

          /\      |‾‾| /‾‾/   /‾‾/   
     /\  /  \     |  |/  /   /  /    
    /  \/    \    |     (   /   ‾‾\  
   /          \   |  |\  \ |  (‾)  | 
  / __________ \  |__| \__\ \_____/ .io

DEBU[0000] Resolving and reading test 'examples/mysql_secure_test.js'... 
DEBU[0000] Loading...                                    moduleSpecifier="file:///<redacted>/xk6-sql/examples/mysql_secure_test.js" originalModuleSpecifier=examples/mysql_secure_test.js
DEBU[0000] 'examples/mysql_secure_test.js' resolved to 'file:///<redacted>/code/xk6-sql/examples/mysql_secure_test.js' and successfully loaded 1141 bytes! 
DEBU[0000] Gathering k6 runtime options...              
DEBU[0000] Initializing k6 runner for 'examples/mysql_secure_test.js' (file:///<redacted>/code/xk6-sql/examples/mysql_secure_test.js)... 
DEBU[0000] Detecting test type for...                    test_path="file:///<redacted>/code/xk6-sql/examples/mysql_secure_test.js"
DEBU[0000] Trying to load as a JS test...                test_path="file:///<redacted>/code/xk6-sql/examples/mysql_secure_test.js"
DEBU[0000] Babel: Transformed                            t=40.119292ms
DEBU[0000] Runner successfully initialized!             
DEBU[0000] Parsing CLI flags...                         
DEBU[0000] Consolidating config layers...               
DEBU[0000] Parsing thresholds and validating config...  
DEBU[0000] Initializing the execution scheduler...      
DEBU[0000] Starting 2 outputs...                         component=output-manager
DEBU[0000] Starting...                                   component=metrics-engine-ingester
DEBU[0000] Started!                                      component=metrics-engine-ingester
     execution: local
        script: examples/mysql_secure_test.js
        output: -

     scenarios: (100.00%) 1 scenario, 1 max VUs, 10m30s max duration (incl. graceful stop):
              * default: 1 iterations for each of 1 VUs (maxDuration: 10m0s, gracefulStop: 30s)

DEBU[0000] Trapping interrupt signals so k6 can handle them gracefully... 
DEBU[0000] Starting emission of VUs and VUsMax metrics... 
DEBU[0000] Start of initialization                       executorsCount=1 neededVUs=1 phase=execution-scheduler-init
DEBU[0000] Starting the REST API server on localhost:6565 
DEBU[0000] Initialized VU #1                             phase=execution-scheduler-init
DEBU[0000] Finished initializing needed VUs, start initializing executors...  phase=execution-scheduler-init
DEBU[0000] Initialized executor default                  phase=execution-scheduler-init
DEBU[0000] Initialization completed                      phase=execution-scheduler-init
DEBU[0000] Start of test run                             executorsCount=1 phase=execution-scheduler-run
DEBU[0000] Running setup()...                           
DEBU[0000] Start all executors...                        phase=execution-scheduler-run
DEBU[0000] Starting executor                             executor=default startTime=0s type=per-vu-iterations
DEBU[0000] Starting executor run...                      executor=per-vu-iterations iterations=1 maxDuration=10m0s scenario=default type=per-vu-iterations vus=1
INFO[0000] key: plugin-name, value: k6-plugin-sql        source=console
INFO[0000] key: plugin-name, value: k6-plugin-sql        source=console
INFO[0000] key: plugin-name, value: k6-plugin-sql        source=console
INFO[0000] key: plugin-name, value: k6-plugin-sql        source=console
INFO[0000] key: plugin-name, value: k6-plugin-sql        source=console
INFO[0000] key: plugin-name, value: k6-plugin-sql        source=console
INFO[0000] key: plugin-name, value: k6-plugin-sql        source=console
INFO[0000] key: plugin-name, value: k6-plugin-sql        source=console
INFO[0000] key: plugin-name, value: k6-plugin-sql        source=console
INFO[0000] key: plugin-name, value: k6-plugin-sql        source=console
INFO[0000] key: plugin-name, value: k6-plugin-sql        source=console
INFO[0000] key: plugin-name, value: k6-plugin-sql        source=console
INFO[0000] key: plugin-name, value: k6-plugin-sql        source=console
INFO[0000] key: plugin-name, value: k6-plugin-sql        source=console
INFO[0000] key: plugin-name, value: k6-plugin-sql        source=console
INFO[0000] key: plugin-name, value: k6-plugin-sql        source=console
INFO[0000] key: plugin-name, value: k6-plugin-sql        source=console
INFO[0000] key: plugin-name, value: k6-plugin-sql        source=console
INFO[0000] key: plugin-name, value: k6-plugin-sql        source=console
INFO[0000] key: plugin-name, value: k6-plugin-sql        source=console
INFO[0000] key: plugin-name, value: k6-plugin-sql        source=console
INFO[0000] key: plugin-name, value: k6-plugin-sql        source=console
INFO[0000] key: plugin-name, value: k6-plugin-sql        source=console
INFO[0000] key: plugin-name, value: k6-plugin-sql        source=console
INFO[0000] key: plugin-name, value: k6-plugin-sql        source=console
INFO[0000] key: plugin-name, value: k6-plugin-sql        source=console
INFO[0000] key: plugin-name, value: k6-plugin-sql        source=console
INFO[0000] key: plugin-name, value: k6-plugin-sql        source=console
INFO[0000] key: plugin-name, value: k6-plugin-sql        source=console
INFO[0000] key: plugin-name, value: k6-plugin-sql        source=console
INFO[0000] key: plugin-name, value: k6-plugin-sql        source=console
DEBU[0000] Regular duration is done, waiting for iterations to gracefully finish  executor=per-vu-iterations gracefulStop=30s scenario=default
DEBU[0000] Executor finished successfully                executor=default startTime=0s type=per-vu-iterations
DEBU[0000] Running teardown()...                        
DEBU[0000] Test finished cleanly                        
DEBU[0000] Stopping vus and vux_max metrics emission...  phase=execution-scheduler-init
DEBU[0000] Metrics emission of VUs and VUsMax metrics stopped 
DEBU[0000] Releasing signal trap...                     
DEBU[0000] Sending usage report...                      
DEBU[0000] Waiting for metrics and traces processing to finish... 
DEBU[0000] Metrics and traces processing finished!      
DEBU[0000] Stopping outputs...                          
DEBU[0000] Stopping 2 outputs...                         component=output-manager
DEBU[0000] Stopping...                                   component=metrics-engine-ingester
DEBU[0000] Stopped!                                      component=metrics-engine-ingester
DEBU[0000] Generating the end-of-test summary...        

     data_received........: 0 B 0 B/s
     data_sent............: 0 B 0 B/s
     iteration_duration...: avg=9.11ms min=6.87µs med=13.57ms max=13.75ms p(90)=13.72ms p(95)=13.74ms
     iterations...........: 1   26.986911/s


running (00m00.0s), 0/1 VUs, 1 complete and 0 interrupted iterations
default ✓ [======================================] 1 VUs  00m00.0s/10m0s  1/1 iters, 1 per VU
DEBU[0000] Usage report sent successfully               
DEBU[0000] Everything has finished, exiting k6 normally! 
```

### CODE REVIEW

File: sql.go [hunk 1]
```diff
@@ -58,22 +100,115 @@ func contains(array []string, element string) bool {
...
+	mysqlTLSConfig := &tls.Config{
+		RootCAs:            rootCAs,
+		Certificates:       clientCerts,
+		MinVersion:         TLSVersions[tlsConfig.MinVersion],
+		InsecureSkipVerify: tlsConfig.InsecureSkipTLSverify,
+	}
+	if err := mysql.RegisterTLSConfig(tlsConfigKey, mysqlTLSConfig); err != nil {
+		return err
+	}
+	return nil
```
Comment (mrevutskyi) [inline sql.go:209]: simply `return mysql.RegisterTLSConfig(tlsConfigKey, mysqlTLSConfig)`

File: sql.go [hunk 2]
```diff
@@ -58,22 +100,115 @@ func contains(array []string, element string) bool {
...
+	rootCAs := x509.NewCertPool()
+	{
+		pem, err := os.ReadFile(tlsConfig.CAcertFile)
+		if err != nil {
+			return err
+		}
+		if ok := rootCAs.AppendCertsFromPEM(pem); !ok {
+			return fmt.Errorf("failed to append PEM")
+		}
+	}
```
Comment (mrevutskyi) [inline sql.go:190]: what do we need the extra `{}` block for?

File: sql.go [hunk 3]
```diff
@@ -58,22 +100,115 @@ func contains(array []string, element string) bool {
...
 // the provided connection string.
-func (*SQL) Open(database string, connectionString string) (*dbsql.DB, error) {
+func (sql *SQL) Open(database string, connectionString string) (*dbsql.DB, error) {
 	supportedDatabases := []string{"mysql", "postgres", "sqlite3", "sqlserver", "azuresql", "clickhouse"}
 	if !contains(supportedDatabases, database) {
 		return nil, fmt.Errorf("database %s is not supported", database)
 	}
 
+	if database == "mysql" && sql.tlsConfig.EnableTLS {
+		tlsConfigKey := "custom"
```
Comment (mrevutskyi) [inline sql.go:154]: make it a constant?

File: examples/mysql_secure_test.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
+import sql from 'k6/x/sql';
+
+sql.loadTLS({
+  enableTLS: true,
```
Comment (mrevutskyi) [inline examples/mysql_secure_test.js:4]: this seems redundant, is there a case when we want to loadTLS with `enableTLS: false`?

File: sql.go [hunk 5]
```diff
@@ -58,22 +100,115 @@ func contains(array []string, element string) bool {
...
+		if err := registerTLS(tlsConfigKey, sql.tlsConfig); err != nil {
+			return nil, err
+		}
```
Comment (mrevutskyi) [inline sql.go:157]: why do it here and not inside `LoadTLS`?

File: sql.go [hunk 6]
```diff
@@ -2,20 +2,39 @@
...
+var TLSVersions map[string]uint16
+
 func init() {
+	// Initialize the TLS versions map.
+	TLSVersions = map[string]uint16{
+		netext.TLS_1_0: tls.VersionTLS10,
+		netext.TLS_1_1: tls.VersionTLS11,
+		netext.TLS_1_2: tls.VersionTLS12,
+		netext.TLS_1_3: tls.VersionTLS13,
+	}
```
[the remark this answers is not in the document]
Comment (surajvshankar) [reply sql.go]: Moved it out of `init`.

File: sql.go (hunk 2 above)
Comment (surajvshankar) [reply sql.go]: Done. Thank you.

File: sql.go (hunk 3 above)
Comment (surajvshankar) [reply sql.go]: Done.

File: examples/mysql_secure_test.js (hunk 4 above)
Comment (surajvshankar) [reply examples/mysql_secure_test.js:4]: As discussed, using this as a sentinel in `Open` to `RegisterTLSConfig`.

**Comment by surajvshankar** [comment]
@szkiba : We are attempting to add TLS support to the mysql driver used by this plugin. Could you please review and provide your feedback?

**Comment by szkiba** [comment]
> @szkiba : We are attempting to add TLS support to the mysql driver used by this plugin. Could you please review and provide your feedback?

@surajvshankar
Of course. Could you fix the linter problem first?

**Comment by szkiba** [comment]
@surajvshankar  still has linter issues
https://github.com/grafana/xk6-sql/actions/runs/10721473919/job/29735069904?pr=71

**Comment by surajvshankar** [comment]
> @surajvshankar still has linter issues https://github.com/grafana/xk6-sql/actions/runs/10721473919/job/29735069904?pr=71

Thank you @szkiba. I have fixed the double import. 

I have ensured that on manually running the linter at my end, it now succeeds:
```
golangci-lint run
WARN [config_reader] The configuration option `linters.govet.check-shadowing` is deprecated. Please enable `shadow` instead, if you are not using `enable-all`. 
WARN The linter 'exportloopref' is deprecated (since v1.60.2) due to: Since Go1.22 (loopvar) this linter is no longer relevant. Replaced by copyloopvar. 
```

**Review by szkiba — APPROVED** [review]
LGTM

**Comment by surajvshankar** [comment]
Ivan, we are pulling the "latest" version/tag of xk6-sql. Thank you for
creating the release: https://github.com/grafana/xk6-sql/releases/tag/v0.4.1.
Could you also update the "latest" tag to point to this release, please?

Regards,
suraj.

**Comment by szkiba** [comment]
Done

**Pull request merged**